Add defaultValue option to data getters (T648641)

mrjohn0011 opened the pull request:
Make difference between an undefined value of the property and no property in the object.

Some widgets such as the [dxCheckBox](https://js.devexpress.com/Documentation/ApiReference/UI_Widgets/dxCheckBox/Configuration/#value) may have an `undefined` value but it's default value when there is no value option specified, should be `false`

mrjohn0011 committed:
Use null value for dxTreeView item with indeterminate selection (T648641)

mrjohn0011 committed:
Fix tests

mrjohn0011 committed:
Add test for data utils

mrjohn0011 committed:
Fix data utils test

mrjohn0011 committed:
Refactoring

mrjohn0011 committed:
Performance optimization

mrjohn0011 committed:
Use `in` operator instead of `hasOwnProperty`

mrjohn0011 committed:
Add more test cases

mrjohn0011 committed:
Fix tests

mrjohn0011 committed:
Fix exception when current is primitive

San4es commented on js/core/utils/data.js:
```diff
@@ -62,12 +62,17 @@ var compileGetter = function(expr) {
+                }
 
-                var next = unwrap(current[path[i]], options);
+                var useDefaultValue = defaultValueExists && typeUtils.isPlainObject(current) && !(path[i] in current),
```
> typeUtils.isPlainObject(current)

What about cases with class instances?

```
class A {  };
assert.equal(GETTER("a")(new A(), { defaultValue: "default" }), "default");
```

AlekseyMartynov replied:
Does this answer @San4es' question?

https://github.com/DevExpress/DevExtreme/blob/cd50829772388821d91974517d382d5c58504b3f/testing/tests/DevExpress.core/utils.data.tests.js#L65-L77

San4es replied:
@AlekseyMartynov it's not like that. As far as I see, it is the test scenario for `path[i] in current` vs. `current.hasOwnProperty(path[i])`.

AlekseyMartynov replied:
Got it.
@mrjohn0011 changed to `isObject` in https://github.com/DevExpress/DevExtreme/pull/5212/commits/368010325748a32bebe6fa6e441dfe75ef7ed295

AlekseyMartynov replied:
Assert added:
https://github.com/DevExpress/DevExtreme/blob/4252cc738584de868d68f674e01866cf85b97d29/testing/tests/DevExpress.core/utils.data.tests.js#L86

San4es commented on js/core/utils/data.js:
```diff
@@ -62,12 +62,17 @@ var compileGetter = function(expr) {
 
-                var next = unwrap(current[path[i]], options);
+                var useDefaultValue = defaultValueExists && typeUtils.isPlainObject(current) && !(path[i] in current),
+                    nextValue = useDefaultValue ? options.defaultValue : current[path[i]],
```
`path[i]` is used twice in this code block.

mrjohn0011 committed:
Refactoring and add test assert

AlekseyMartynov committed:
Update data.js

AlekseyMartynov committed:
Update data.js

AlekseyMartynov commented:
@mrjohn0011 
What about this?

```
GETTER("a.b.c")({ }, { defaultValue: { b: 1 } });
```

Do you expect `undefined` or `{ b: 1 }`?

😈

mrjohn0011 commented:
@AlekseyMartynov, I expected { b: 1 }

mrjohn0011 committed:
Add one more case

AlekseyMartynov committed:
Update utils.data.tests.js

AlekseyMartynov committed:
Update data.js

AlekseyMartynov commented on testing/tests/DevExpress.core/utils.data.tests.js:
```diff
@@ -55,10 +55,29 @@ QUnit.test("it works", function(assert) {
     assert.equal(GETTER("c.a.a")(obj), "c.a.a");
     assert.equal(GETTER("d.a")(obj), "d().a");
     assert.equal(GETTER("e.z")(obj), undefined);
+    assert.equal(GETTER("a")(null), null);
```
Not sure this is/was correct 🤔

mrjohn0011 committed:
Add more test cases

mrjohn0011 committed:
Separate test cases

mrjohn0011 committed:
Update

mrjohn0011 committed:
Take out the trash

mrjohn0011 committed:
Update

AlekseyMartynov committed:
Update utils.data.tests.js

- combine with `inheritance`
- check reading from `undefined`
- check reading from a non-plain object (suggested by Zi)

AlekseyMartynov reviewed (approved):
🥇
